Fix MemberExpression and add MethodCalls

jonsmock opened the pull request:
Builds on https://github.com/Viasat/dctest/pull/15, only last 3 commits are new.

Fixes a bug in previous member access to allow things like `fromJSON(...).foo`.  Previously only `foo.a.b.c` would work, but you could not access members on values returned by functions.

Adds MethodCall to the ast to support methods like `x.toString()` and `x.count()`.

Adds some additional functions for strings and conversions.

jonsmock committed:
Add initial string and JSON conversion functions

Upgrade cljs-bean when using shadow-cljs to get :keywordize-keys
functionality.

jonsmock committed:
Allow member access for results of function calls or parens expressions

Previously MemberExpression prohibited something like `foo(env).b.c`,
because `foo(env)` is a FunctionCall, not a ContextName.

We now require that MemberExpression contains at least one Property
access. In order to match `foo(env).b.c` correctly, we must move
MemberExpression to the first matching position of UnaryExpression;
otherwise, FunctionCall will match `foo(env)`, followed by an end of
input error when attempting to match `.b.c`.

Choosing for now to allow member access on Value, but that is up for
debate (ex: `{"a": 3}.a`).

jonsmock committed:
Support method calls

Initial method call support with the first supported methods including
`toString` and `count`.

MemberExpression supports both MethodCall and Property, so that we can
greedily parse expressions like `x.foo().bar.baz().z`. If we attempt to
separate methods and properties into separate expression types, the
MemberExpression will attempt to capture `foo` and `baz` as member
accesses, and we run into end of input errors for the method calls.

MemberExpression evaluation needs to also evaluate its MethodCall
children, so we can pass the referenced object of the MethodCall. This
works but has the drawback of needing to check for uncaught errors again
inside the MemberExpression evaluation.

kanaka reviewed (changes requested):
I think the functions `fromJSON`, `toJSON`, `contains`, etc, are used before the commit where they are defined.

Otherwise, looks good.

jonsmock commented:
Gah, thanks!  I've moved the offending examples/tests to the last commit, where those functions are introduced (and fixed some typos in the commit messages).
